sdk/keyvault/azure-security-keyvault-keys/src/cryptography/cryptography_client.cpp

```diff
@@ -0,0 +1,389 @@
+    {
+      // If provider supports remote, otherwise re-throw
+      if (!m_provider->CanRemote())
+      {
+        throw;
+      }
+    }
+
+    if (result.Ciphertext.empty())
+    {
+      ThrowIfLocalOnly(KeyOperation::Encrypt.ToString());
+
+      result = m_remoteProvider->Encrypt(parameters, context);
+    }
+  }
+
+  return result;
+}
+
+DecryptResult CryptographyClient::Decrypt(
+    DecryptParameters const& parameters,
+    Azure::Core::Context const& context)
+{
+  if (m_provider == nullptr)
+  {
+    // Try to init a local crypto provider after getting the key from the server.
+    // If the local provider can't be created, the remote client is used as provider.
+    Initialize(KeyOperation::Decrypt.ToString(), context);
+  }
+
+  // Default result has empty values.
+  DecryptResult result;
+
+  // m_provider can be local or remote, depending on how it was init.
+  if (m_provider->SupportsOperation(KeyOperation::Decrypt))
+  {
+    try
+    {
+      result = m_provider->Decrypt(parameters, context);
+    }
+    catch (std::exception const&)
+    {
+      // If provider supports remote, otherwise re-throw
+      if (!m_provider->CanRemote())
+      {
+        throw;
+      }
+    }
+
+    if (result.Plaintext.empty())
+    {
+      ThrowIfLocalOnly(KeyOperation::Decrypt.ToString());
+
+      result = m_remoteProvider->Decrypt(parameters, context);
+    }
+  }
+
+  return result;
+}
+
+WrapResult CryptographyClient::WrapKey(
+    KeyWrapAlgorithm algorithm,
+    std::vector<uint8_t> const& key,
+    Azure::Core::Context const& context)
+{
+  if (m_provider == nullptr)
+  {
+    // Try to init a local crypto provider after getting the key from the server.
+    // If the local provider can't be created, the remote client is used as provider.
+    Initialize(KeyOperation::WrapKey.ToString(), context);
+  }
+
+  // Default result has empty values.
+  WrapResult result;
+
+  // m_provider can be local or remote, depending on how it was init.
+  if (m_provider->SupportsOperation(KeyOperation::WrapKey))
+  {
+    try
+    {
+      result = m_provider->WrapKey(algorithm, key, context);
+    }
+    catch (std::exception const&)
+    {
+      // If provider supports remote, otherwise re-throw
+      if (!m_provider->CanRemote())
+      {
+        throw;
+      }
+    }
+
+    if (result.EncryptedKey.size() == 0)
+    {
+      ThrowIfLocalOnly(KeyOperation::WrapKey.ToString());
+
+      result = m_remoteProvider->WrapKey(algorithm, key, context);
+    }
+  }
+
+  return result;
+}
+
+UnwrapResult CryptographyClient::UnwrapKey(
+    KeyWrapAlgorithm algorithm,
+    std::vector<uint8_t> const& encryptedKey,
+    Azure::Core::Context const& context)
+{
+  if (m_provider == nullptr)
+  {
+    // Try to init a local crypto provider after getting the encryptedKey from the server.
+    // If the local provider can't be created, the remote client is used as provider.
+    Initialize(KeyOperation::UnwrapKey.ToString(), context);
+  }
```

Comment:
For what I learned from @heaths, there's never been a case where some client ask for a remote-only client.  The SDK gives preference to local crypto operations leveraging the available language-crypto-libraries.  And the Key Vault service is used as a plan B in case the local provider can't do the job.

In this PR, I implement all the required machinery to follow the same approach as .NET and other languages. However, I haven't added all the support for local crypto providers for C++ (openssl in linux and bcrypt on Windows), so all the calls are currently handled by the server.

Later, once we implement the local providers, the behavior will be 100% equal to the other languages